test/core/security/grpc_audit_logging_test.cc

```diff
@@ -81,26 +85,87 @@ class AuditLoggingTest : public ::testing::Test {
 
 }  // namespace
 
-TEST_F(AuditLoggingTest, SuccessfulLoggerCreation) {
+//
+// AuditLoggerRegistryTest
+//
+
+TEST_F(AuditLoggerRegistryTest, SuccessfulLoggerCreation) {
   auto result = AuditLoggerRegistry::ParseConfig(kName, Json());
   ASSERT_TRUE(result.ok());
   ASSERT_NE(AuditLoggerRegistry::CreateAuditLogger(std::move(result.value())),
             nullptr);
 }
 
-TEST_F(AuditLoggingTest, UnknownLogger) {
+TEST_F(AuditLoggerRegistryTest, UnknownLogger) {
   auto result = AuditLoggerRegistry::ParseConfig("unknown_logger", Json());
   EXPECT_EQ(result.status().code(), absl::StatusCode::kNotFound);
   EXPECT_EQ(result.status().message(),
             "audit logger factory for unknown_logger does not exist")
       << result.status();
 }
 
-TEST_F(AuditLoggingTest, AuditLoggerFactoryExistenceChecks) {
+TEST_F(AuditLoggerRegistryTest, LoggerFactoryExistenceChecks) {
   EXPECT_TRUE(AuditLoggerRegistry::FactoryExists(kName));
   EXPECT_FALSE(AuditLoggerRegistry::FactoryExists("unknown_logger"));
 }
 
+//
+//  StdoutLoggerTest
+//
+
+TEST(StdoutLoggerTest, LoggerFactoryExistenceChecks) {
+  EXPECT_TRUE(AuditLoggerRegistry::FactoryExists("stdout_logger"));
+}
+
+TEST(StdoutLoggerTest, BadLoggerConfig) {
+  auto invalid_value = Json::FromBool(false);
+  auto result =
+      AuditLoggerRegistry::ParseConfig("stdout_logger", invalid_value);
+  ASSERT_EQ(result.status().code(), absl::StatusCode::kInvalidArgument);
+  ASSERT_EQ(result.status().message(), "config is not a json object")
+      << result.status();
+}
+
+TEST(StdoutLoggerTest, StdoutLoggerCreationAndLogInvocation) {
+  auto result =
+      AuditLoggerRegistry::ParseConfig("stdout_logger", Json::FromObject({}));
+  ASSERT_TRUE(result.ok());
+  auto logger =
+      AuditLoggerRegistry::CreateAuditLogger(std::move(result.value()));
+  AuditContext context("method", "spiffe", "policy", "rule", true);
+  ::testing::internal::CaptureStdout();
+  logger->Log(context);
+  auto log_or = JsonParse(::testing::internal::GetCapturedStdout());
+  ASSERT_TRUE(log_or.ok());
+  ASSERT_EQ(log_or.value().type(), Json::Type::kObject);
+  auto it = log_or.value().object().find("grpc_audit_log");
+  ASSERT_NE(it, log_or.value().object().end());
+  ASSERT_EQ(it->second.type(), Json::Type::kObject);
+  auto object = it->second.object();
+  // Check if the recorded timestamp is close to now.
+  absl::Time t1 = absl::Now();
+  ASSERT_NE(object.find("timestamp"), object.end());
+  EXPECT_EQ(object.find("timestamp")->second.type(), Json::Type::kNumber);
+  uint64_t ts;
+  ASSERT_TRUE(absl::SimpleAtoi(object.find("timestamp")->second.string(), &ts));
+  absl::Time t2 = absl::FromUnixSeconds(ts);
+  EXPECT_TRUE(t1 - t2 < absl::Seconds(10));
```

Comment:
Rather than hard-coding this 10-second tolerance, I suggest storing the time before we call `logger->Log()`, then storing the time after that call returns, and then verifying that the logged timestamp is between those two values.

Reply:
Done.